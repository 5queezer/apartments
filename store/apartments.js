import api from './api.kasaz'

export const state = () => ({
  list: [],
  id: undefined
})

export const mutations = {
  add (state, apartment) {
    if (typeof apartment.address.lat === 'undefined' || typeof apartment.address.long === 'undefined') {
      throw new TypeError(`malformed geo location { ${apartment.lat} ${apartment.long} }`)
    }
    if (typeof apartment.id === 'undefined') {
      apartment.id = Math.max(state.list.map(s => s.id)) + 1
    } else {
      for (const entry of state.list) {
        if (entry.id === apartment.id) { throw new Error('double entry') }
      }
    }

    state.list.push(apartment)
  },
  set (state, id) {
    state.id = id
  }
}

export const actions = {
  fetchAll ({ commit }) {
    return new Promise(async (resolve, reject) => {
      try {
        const params = {
          'location[city]': 'Barcelona',
          'filters[price][min]': 1.5e6
        }
        const apartments = await api.getApartments(params)
        if (!apartments) { reject(new Error('api.getApartments() result is empty')) }
        for (const a of apartments) { commit('add', a) }
        resolve(apartments.length)
      } catch (error) {
        reject(error)
      }
    })
  }
}

export const getters = {
  list (state) {
    return state.list
  },
  current (state) {
    return state.list.find(apartment => apartment.id)
  }
}
