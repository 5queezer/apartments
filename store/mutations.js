export default {
  SET_APARTMENT (state, apartment) {
    for (const a of state.apartments) {
      if (a.id === apartment.id) { throw new Error('double entry') }
    }
    state.apartments.push({
      id: apartment.id,
      title: apartment.title,
      price: apartment.price,
      bedrooms: apartment.bedrooms,
      bathrooms: apartment.bathrooms,
      picture: apartment.picture,
      address: {
        city: apartment.address.city,
        zip: apartment.address.zip,
        street: apartment.address.street,
        lat: apartment.address.lat,
        long: apartment.address.long
      }
    })
  },
  SET_APARTMENTS (state, apartments) {
    for (const a of apartments) { this.SET_APARTMENT(a) }
  }
}
