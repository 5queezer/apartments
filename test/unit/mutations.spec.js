import apartments from './fixtures/apartments'
import mutations from '@/store/mutations'
import initialState from '@/store/state'

describe('Mutations', () => {
  let state
  const apartment = apartments[0]

  beforeEach(() => {
    state = { ...initialState }
  })

  it('adds a new apartment', () => {
    // arrange
    const expectedApartment = {
      id: 0,
      title: apartment.title,
      price: apartment.price,
      bedrooms: apartment.bedrooms,
      bathrooms: apartment.bathrooms,
      picture: apartment.picture,
      address: {
        city: apartment.address.city,
        zip: apartment.address.zip,
        street: apartment.address.street,
        lat: apartment.address.lat,
        long: apartment.address.long
      }
    }

    // act
    mutations.SET_APARTMENT(state, apartment)

    // assert
    expect(state.apartments.length).toBe(1)
    expect(state.apartments[0]).toEqual(expectedApartment)
  })
})
