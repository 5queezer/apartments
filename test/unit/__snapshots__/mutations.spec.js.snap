// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutations throws error with entry of same id 1`] = `"double entry"`;
